modules/rostests/apitests/shell32/OpenAs_RunDLL.cpp

```diff
@@ -0,0 +1,192 @@
+// OpenAs_RunDLLW
+typedef void (WINAPI *OPENAS_RUNDLLW)(HWND, HINSTANCE, LPCWSTR, int);
+
+static OPENAS_RUNDLLA pOpenAs_RunDLLA = NULL;
+static OPENAS_RUNDLLW pOpenAs_RunDLLW = NULL;
+
+struct TEST_ENTRY
+{
+    int         nLine;
+    BOOL        bWide;
+    HINSTANCE   hInst;
+    int         nRet;
+    BOOL        bCreateFile;
+    LPCSTR      pszFileA;
+    LPCWSTR     pszFileW;
+};
+
+#define COUNT       5
+#define INTERVAL    200
+
+#define DEATH       999
+#define OPENED      1
+#define NOT_OPENED  0
+
+#define BAD_INST   ((HINSTANCE)0xDEAD)
+#define BAD_SZ_A   ((LPCSTR)0xDEAD)
+#define BAD_SZ_W   ((LPCWSTR)0xDEAD)
+
+static const TEST_ENTRY s_TestEntries[] =
+{
+    // ANSI
+    {__LINE__, FALSE, NULL, OPENED, FALSE, NULL, NULL },
+    {__LINE__, FALSE, NULL, OPENED, FALSE, "", NULL },
+    {__LINE__, FALSE, NULL, OPENED, FALSE, "invalid file name.txt", NULL },
+    {__LINE__, FALSE, NULL, DEATH, FALSE, BAD_SZ_A, NULL },
+    {__LINE__, FALSE, NULL, OPENED, TRUE, "created file.txt", NULL },
+    {__LINE__, FALSE, BAD_INST, OPENED, FALSE, NULL, NULL },
+    {__LINE__, FALSE, BAD_INST, OPENED, FALSE, "invalid file name.txt", NULL },
+    {__LINE__, FALSE, BAD_INST, DEATH, FALSE, BAD_SZ_A, NULL },
+    {__LINE__, FALSE, BAD_INST, OPENED, TRUE, "created file.txt", NULL },
+    // WIDE
+    {__LINE__, TRUE, NULL, DEATH, FALSE, NULL, NULL },
+    {__LINE__, TRUE, NULL, OPENED, FALSE, NULL, L"" },
+    {__LINE__, TRUE, NULL, OPENED, FALSE, NULL, L"invalid file name.txt" },
+    {__LINE__, TRUE, NULL, DEATH, FALSE, NULL, BAD_SZ_W },
+    {__LINE__, TRUE, NULL, OPENED, TRUE, NULL, L"created file.txt" },
+    {__LINE__, TRUE, BAD_INST, DEATH, FALSE, NULL, NULL },
+    {__LINE__, TRUE, BAD_INST, OPENED, FALSE, NULL, L"invalid file name.txt" },
+    {__LINE__, TRUE, BAD_INST, DEATH, FALSE, NULL, BAD_SZ_W },
+    {__LINE__, TRUE, BAD_INST, OPENED, TRUE, NULL, L"created file.txt" },
+};
+
+HANDLE g_hThread = NULL;
+INT g_nRet = 0;
+
+static unsigned __stdcall
+watch_thread_proc(void *arg)
+{
+    for (int i = 0; i < COUNT; ++i)
+    {
+        Sleep(INTERVAL);
+
+        HWND hwnd = FindWindowA("#32770", "Open With");
+        if (hwnd == NULL)
+            continue;
+
+        g_nRet = OPENED;
+        PostMessage(hwnd, WM_COMMAND, MAKEWPARAM(IDCANCEL, BN_CLICKED), 0);
+        break;
+    }
+    return 0;
+}
+
+static void StartWatchGUI(const TEST_ENTRY *pEntry)
+{
+    g_nRet = NOT_OPENED;
+    g_hThread = (HANDLE)_beginthreadex(NULL, 0, watch_thread_proc, NULL, 0, NULL);
+}
+
+static void DoEntry(const TEST_ENTRY *pEntry)
+{
+    _SEH2_TRY
+    {
+        if (pEntry->bWide)
+        {
+            if (pEntry->bCreateFile)
+            {
+                FILE *fp = _wfopen(pEntry->pszFileW, L"wb");
+                ok(fp != NULL, "Line %d: Unable to create file '%s'\n", pEntry->nLine, wine_dbgstr_w(pEntry->pszFileW));
+                fclose(fp);
+            }
+
+            StartWatchGUI(pEntry);
+            SetLastError(0xDEADFACE);
```

Comment:
You're setting last error here but never checking it after.

Reply:
Checked. OK.